sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1174 @@
+            raise ValueError(
+                f"Given callee:{callee} is not a valid method. Valid methods are:"
+                f" {METHODS}"
+            )
+        if caller not in METHODS:
+            raise ValueError(
+                f"Given caller:{caller} is not a valid method. Valid methods are:"
+                f" {METHODS}"
+            )
+        self._routes.append(MethodPair(callee=callee, caller=caller))
+        return self
+
+    def _serialize(self):
+        """Serialize the object.
+
+        Returns
+        -------
+        obj : list
+            A serialized version of the instance in the form of a list.
+        """
+        result = list()
+        for route in self._routes:
+            result.append({"callee": route.callee, "caller": route.caller})
+        return result
+
+    @classmethod
+    def from_str(cls, route):
+        """Construct an instance from a string.
+
+        Parameters
+        ----------
+        route : str
+            A string representing the mapping, it can be:
+
+              - `"one-to-one"`: a one to one mapping for all methods.
+              - `"method"`: the name of a single method.
+
+        Returns
+        -------
+        obj : MethodMapping
+            A :class:`~utils.metadata_requests.MethodMapping` instance
+            constructed from the given string.
+        """
+        routing = cls()
+        if route == "one-to-one":
+            for method in METHODS:
+                routing.add(callee=method, caller=method)
+        elif route in METHODS:
+            routing.add(callee=route, caller=route)
+        else:
+            raise ValueError("route should be 'one-to-one' or a single method!")
+        return routing
+
+    def __repr__(self):
+        return str(self._serialize())
+
+    def __str__(self):
+        return str(repr(self))
+
+
+class MetadataRouter:
+    """Stores and handles metadata routing for a router object.
+
+    This class is used by router objects to store and handle metadata routing.
+    Routing information is stored as a dictionary of the form ``{"object_name":
+    RouteMappingPair(method_mapping, routing_info)}``, where ``method_mapping``
+    is an instance of :class:`~utils.metadata_requests.MethodMapping` and
+    ``routing_info`` is either a
+    :class:`~utils.metadata_requests.MetadataRequest` or a
+    :class:`~utils.metadata_requests.MetadataRouter` instance.
+
+    .. versionadded:: 1.1
+
+    Parameters
+    ----------
+    owner : str
+        The name of the object to which these requests belong.
+    """
+
+    # this is here for us to use this attribute's value instead of doing
+    # `isinstance`` in our checks, so that we avoid issues when people vendor
+    # this file instad of using it directly from scikit-learn.
+    type = "metadata_router"
+
+    def __init__(self, owner):
+        self._route_mappings = dict()
+        self._self = None
+        self.owner = owner
+
+    def add_self(self, obj):
```

Comment:
I found this very confusing. IIUK, it adds `obj` so `_self` and `_self` is an attribute with special treatment/meaning in routing.

Reply:
Tried adding some clarification here and above.